### PR TITLE
feat(#715): Phase 3 slice 1 — root implicit AD on SymmetricTensor

### PR DESCRIPTION
**Pull request by yingjerkao**
> 🤖 **AI-generated PR** — written by Claude Code, posted by @yingjerkao.

Ports root implicit AD ([arXiv:2607.15030](https://arxiv.org/abs/2607.15030)) to
block-sparse `SymmetricTensor` — #715 Phase 3, slice 1. Bosonic abelian, 1×1 unit
cell, D=2, χ=4.

Design: `docs/plans/2026-07-31-715-phase3-symmetric-design.md`
Plan: `docs/superpowers/plans/2026-07-31-715-phase3-symmetric-root-implicit.md`

## Result

| quantity | value |
|---|---|
| symmetric gradient vs dense root-implicit | **3.29e-15** relative |
| energy vs dense | 1.5e-16 |
| `norm(F(y*))` | 2.70e-13 |
| `gmres_residual` | 6.38e-15 |
| directional FD (sweep map, 12/20/30 sweeps) | 4.28e-4 / 7.67e-7 / 5.14e-9 |
| forward energy vs dense | 1.7e-16, same 36 iterations |
| `svd`/`eigh` primitives in the graph | **none** |

The FD ladder converges *onto* the implicit gradient, which is the signature of a
correct implicit gradient against a truncated reference rather than agreement by
construction.

## What this buys, stated precisely

**The qualitative claim is demonstrated.** No `svd` and no `eigh` primitive appears
anywhere in the differentiated graph — asserted over six jaxprs covering the forward
functions *and* their pullbacks, since a decomposition's VJP contains no decomposition
primitive and a pullback-only gate would be hollow. So the degenerate-SV failure class
(the `_gauge_fixed_svd` NaN cluster, `projector_backward="lorentzian"`, #687's accuracy
floor) is structurally absent, not mitigated.

Sharper than #715 anticipated: explicit backprop through the symmetric sweep does not
merely NaN, it **raises** — `sector_svd` ranks singular values across sectors in Python,
so `jax.grad` dies with `ConcretizationTypeError` at D=2. On this path the implicit
gradient is not an improvement over explicit backprop; it is the only thing that
produces a symmetric gradient at all.

**The quantitative claim is not.** Peak RSS is **8.4 GB**, essentially all in the GMRES
solve where XLA compiles the ~15k-equation block-sparse VJP inside a `lax.while_loop`
(0.69 GB after root extraction → 2.33 after the energy VJP → 2.50 after the `F` VJP
trace → 8.36 after the solve). `restart=10` reaches only 7.19 GB while being slower
(93 s vs 49 s) and five orders less accurate. **Removing the SVD VJP did not remove the
compile-scale problem — it moved it.** Whether that is a net win against
`_jit_fused_fixed_point_bwd` is the follow-up measurement, which should now track memory
alongside op count.

## Two deviations from the issue text

- **The `kron` is deleted, not ported.** `jnp.kron(root, eye_d2)` is not a per-sector
  kron once the cut leg is fused: sector `q` mixes every `(q_chi, q_d2)` with
  `q_chi + q_d2 = q`. The leg stays split and `s` is applied by contraction; fusion
  happens only to feed the SVD.
- **`U_perp` is materialised.** #715 planned to avoid it via `(1 - U*U*†)w`; per sector
  the matrices are tiny, and the complementary form adds a redundant null direction to
  `∂_yF` for nothing.

## Structure

Two modules, split so they fail differently — a sector-layer bug is a wrong shape and
raises; a CTM-layer bug is a mis-glued network and returns a plausible wrong number.

- `_ctm_root_implicit_sym_sectors.py` — charge arithmetic, no CTM knowledge.
- `_ctm_root_implicit_symmetric.py` — network, no charge arithmetic.

## Gates

Forward energy vs dense; `norm(F(y*))`; gradient vs the dense FD-validated
root-implicit path; directional FD; a jaxpr assertion; and counterfactuals proving the
residual is load-bearing (dropping the enlarged-corner transpose → 20.8, an extra bridge
shift → 19.4, real-projecting λ → 4.87).

**The layout trap tests.** `norm(F(y*))` being small does not by itself validate the
charge bookkeeping, so two tests make the sector layout load-bearing:

- Moving one retained dimension between sectors (χ preserved) turns out to be *not even
  expressible*: the layout is recorded in the environment itself, as index
  multiplicities on every corner and edge χ leg, so re-extraction is refused outright.
  Forced consistently through a `layout_override` hook, `‖F‖` floors at **4.0e-1** —
  twelve orders above 2.7e-13, and identical at 3 and at 40 polish steps, so it is not a
  matter of polishing further. The same hook with the *right* layout reproduces
  2.697e-13 exactly, so the hook itself is inert.
- A companion trap covers the case no guard can catch: at χ=4 both Z2 sectors have
  literally identical block shapes, so exchanging their blocks passes every shape check,
  charge check and bond index. `‖F‖` → **1.19e3** (27.0 for `S` alone, 39.9 for the
  isometries alone), so the sector assignment is live in both halves of the equations.

Coverage spans Z2 (equal sectors) and U(1) (unequal, fragmenting). The U(1) fixture is
covered by **sweep-by-sweep parity** rather than a converged limit, because it does not
converge — and neither does it in the dense FD-validated module (112 variants searched,
zero converged). That is a property of the state, not the port.

Testing both is not belt-and-braces: three separate bugs were caught only because a
fixture had *unequal* sector sizes. `sector_svd` never oriented its blocks, so `U` and
`Vh` silently swapped — invisible while every test sector was square. `v`'s null-space
shape was read off `U.shape[0]` where it needs `Vh.shape[1]`. And the block-key charge
rule only held for an IN row leg paired with an OUT col leg — `._validate()` passes on
wrong data there, because flow-weighted charges cancel for *any* `q`; only a numeric
round-trip catches it.

One coverage gap closed late: every U(1) projector-vs-dense comparison had been taken on
the *initialiser's* environment. Added a comparison on **swept** U(1) environments
(6 sweeps × 4 directions, worst relative diff **2.47e-14**), where the layout genuinely
moves — `{-1:1, 0:2, 1:1}` → `{0:3, 1:1}` → `{-1:1, 0:3}` → `{0:4}` on one direction
while others hold, so one test covers three-sector and single-sector cuts, sectors
emptying and growing, and four directions disagreeing. No convergence is asserted.

## CI note — please read, this touches shared test infrastructure

`conftest.py` now lets an explicit `@pytest.mark.slow` *withhold* a file's auto-applied
`core` marker. Without it an 8.4 GB test sits in the required `pytest -m core` gate
against GitHub's ~7 GB Linux runners. `-m core` is the only positive selector, so `core`
is the only bucket where a file default can silently override an explicit marker; the
two non-core buckets already select `not core and not slow`, so `algorithm` files are
deliberately untouched.

The marker change alone would **not** have been enough: the 8.4 GB object is a
module-scoped fixture, and four of its consumers were unmarked, so the first to run
would still have paid the full cost. All six consumers are now marked.

Measured with `/usr/bin/time -v`:

| | peak RSS | |
|---|---|---|
| this file under `-m core` | 8.4 GB → **3.62 GB** | 48 passed |
| this file under `-m slow` | 8.81 GB | 6 passed |
| whole `-m core` suite | **6.35 GB** | 1151 passed, 8 skipped, 0 failed |

Collection deltas: `-m core` 1161 → 1159, `-m slow` 70 → 74, `-m algorithm` 907 → 907
(bit-for-bit unchanged), `-m "core and slow"` 2 → 0.

**Two residual risks, neither introduced here, both worth their own issue:**
1. The 8.8 GB is now the `slow` bucket's problem. `test-full` also runs on
   `ubuntu-latest`. It is not a required check and only runs on push-to-`main` or with
   the `run-full-tests` label, so it does not gate this PR — but the OOM risk was
   *relocated, not removed*.
2. `-m core` peaks at 6.35 GB against a ~7 GB runner (~0.65 GB headroom). Pre-existing —
   this change lowers it — but the conftest's own tuning note records 5.6 GB for a
   735-test core bucket, and the bucket is now 1159 tests.

Separately: `test_ctm_root_implicit_multisite.py` (Phase 2, #728) is not in
`_FILE_MARKERS`, so it collects zero tests under `-m core` and has never run in the
required gate. Pre-existing on `main` and left alone here to keep this branch focused.

## Not in this slice

Fermionic `FermionParity`; multisite symmetric (blocked on #667); production wiring
(needs a single half-infinite SVD recipe, #715 obstacle 1); and the #566/#687
measurement. Blockers are named in the design doc §8.

### CODE REVIEW

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Follow-ups from this PR are filed and **deliberately not fixed here**, to keep the branch focused:

- #730 — `test_ctm_root_implicit_multisite.py` (Phase 2, #728) is absent from `_FILE_MARKERS`, so it has never run in a required check. This PR fixes the same omission for its own two files.
- #731 — the adjoint peaks at 8.4 GB in the GMRES solve. Kept out of the `core` gate here, but relocated rather than removed, and it qualifies the #566 cost argument: removing the SVD VJP moved the compile-scale problem rather than eliminating it.
- #732 — `pytest -m core` peaks at 6.35 GB against ~7 GB runners. Pre-existing; this PR lowers it, but the headroom is thin and nothing measures it.

**Comment by yingjerkao** [comment]
@codex review

> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Re-review requested: the previous pass reviewed `a8338a9`, and three commits have landed since.

Your P2 on `_same_block_structure` was **correct and is fixed** in `d00042f`. Reproduced before fixing — the predicate returned `True` for two tensors with block shapes `[(1,1),(2,2)]` and `[(2,2),(1,1)]`, identical key set `((0,0),(1,1))` and identical packed length (`1²+2² == 2²+1²`), so `_max_abs_diff_sym` would have subtracted mismatched flat-buffer entries. A regression test constructs that exact trap.

One deliberate divergence from the suggestion, worth your eye: you proposed including `_block_shapes` (and preferably the index metadata). I implemented **only the index metadata** — per-leg `(sectors, multiplicities, flow)` — on the reasoning that the index layout determines the packing and block shapes are derived from it, so comparing both would be two sources of truth for one fact. If you think block shapes can diverge from the index layout in a way that matters here, that is worth flagging.

Also landed since your pass, both test-only and both platform-calibration rather than logic:
- `ad1d78c` — gates the swept-U(1) projector parity on the truncation-boundary gap. It was failing on all four CI jobs at a boundary where two singular values in different charge sectors are bit-identical, so "the top chi" is genuinely ambiguous and different LAPACK builds retain different subspaces. Tolerance stays 1e-10; a minimum comparison count keeps it from going vacuous.
- `a8338a9` — makes five counterfactual thresholds scale-free (absolute floor plus a ratio against the true residual) after `assert wrong > 1.0` measured 0.924 on macOS and 4.9 on Linux.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Both P2s from the `d00042f` pass were real. Verified each independently before touching
anything; one of them was more consequential than the review states. Fixed in `59322f1`
and `b9bab44`.

## P2 — "Preserve layout overrides during polish sweeps" ✅ correct, and it invalidated a claim

Confirmed: `root_parametrize_sym` forwarded `layout_override` to `all_projectors_sym` and to
the sweep in the layout-mismatch branch, but dropped it at line 1704 — the sweep that
advances the environment at the end of each polish step. The run alternated forced and
natural truncation exactly as described.

The reviewer's inference about the consequence was right, and it goes further than "the
residual cannot establish the claim" — **the claim is false.** Forwarded consistently, the
forced layout converges to a root of its own:

| polish sweeps | 3 | 10 | 40 | converged from a fresh env |
|---|---|---|---|---|
| `norm(F)` | 8.8e-2 | 2.4e-6 | 9.3e-11 | **2.2e-14** |

against 2.7e-13 for the natural layout. The old 4.0e-1 "floor" was the floor of the
alternation. Physically the new number is the right answer: fixing the retained charge
distribution poses a *different* truncation problem, and that problem has a fixed point.

So the trap test was asserting the wrong thing. It now converges the forced layout on its
own terms, asserts it **is** a root (`< 1e-8`), and discriminates on the state instead —
`E = -0.10873` forced vs `-0.10502` natural, a 3.7e-3 gap next to the 1e-10 the natural
layout reproduces against the dense module. The layout is still load-bearing; `norm(F)`
just isn't what detects it. Withdrawal recorded in the design doc §6 and in the test
docstring.

## P2 — "Derive reconstructed keys through the symmetry group" ✅ correct mechanism, different blast radius

The mechanism is exactly as described, and I measured the example:

```
encode(1,2) = 131073      decode(-131073) = (-1, -3)     # raw negation
dual(131073) = -65537     decode(-65537)  = (-1, -2)     # correct
fuse(q, -q) = -65536                                     # not the identity
```

Two things the review didn't separate, though, and they need different treatment:

**`ProductSymmetry` — genuinely broken, now refused.** Conservation is actually violated and
`_from_blocks_unchecked` carries it through. Took the review's own fallback ("or explicitly
reject product symmetries") rather than the primary suggestion, because a correct
`dual`/`fuse` reconstruction is a larger change than this parity-gated slice should carry.

**`Z_n` — reachable *today*, and deliberately left alone.** This is the part I'd flag back:
the first symmetry this bites is not the product group, it's `Z_n`, which every fixture in
this PR uses. A Z2 partner is written `-1` where the canonical representative is `1`. It
survives because `fuse` reduces mod `n`, so `_validate()` passes and the labels are
self-consistent.

But it is **not this module's invention** — the root cause is upstream, in
`_group_blocks_by_bond_charge`:

```
ZnSymmetry(2).fuse_many([[-1]])        -> [-1]      # one array: no % n
ZnSymmetry(2).fuse_many([[-1], [0]])   -> [1]       # two: reduced
```

so `tenax.linalg.svd` on a Z2 matrix emits bond sectors `[0,1]` with the left leg IN and
`[-1,0]` with it OUT. Same tensor, same decomposition. This module's cut builds both
projectors with row and col flowing the same way, which is why it lands in the second row.

Canonicalising here alone would make these bonds fail to contract against the rest of the
library, so the convention is **pinned by a test** that should fail and be updated in step
when the upstream fix lands, rather than left to drift. Filed as #733 with the 7 affected
`linalg.py` call sites.

## Also in this push

`test_zn_round_trips_when_the_flows_are_opposite`,
`test_zn_same_flow_bonds_carry_the_library_s_own_representative`, and
`test_product_symmetry_is_refused_rather_than_mis_assembled` — the last measures its own
premise rather than asserting it from a docstring.

75 tests pass across both Phase 3 files.

On the earlier `_same_block_structure` divergence I raised: still comparing index metadata
only, not `_block_shapes`. No pushback arrived, and I couldn't construct a case where block
shapes diverge from the index layout, since the layout determines the packing — but that
remains the weakest of my assumptions here, so it's still open to challenge.

File: src/tenax/algorithms/_ctm_root_implicit_symmetric.py [hunk 1]
```diff
@@ -0,0 +1,2069 @@
...
+def _same_block_structure(x: SymmetricTensor, y: SymmetricTensor) -> bool:
+    """Whether two symmetric tensors can be compared entry by entry.
+
+    The dense module skips a sweep pair "whose shapes changed".  Here the shape
+    is not the whole story: the retained charge distribution over the cut is
+    data-dependent, so two sweeps can produce the same total ``chi`` split
+    differently over sectors — same dense shape, different blocks.  Comparing
+    the flat buffers in that case would subtract unrelated entries.
+    """
+    return x._data.shape == y._data.shape and x._block_keys == y._block_keys
```
[the remark this answers is not in the document]
Comment (yingjerkao) [reply src/tenax/algorithms/_ctm_root_implicit_symmetric.py]: > 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

All three carried to **#800**. Two are marked outdated by GitHub and I did not re-verify them against current `main`, so the issue says so explicitly rather than implying otherwise.

Singling out the `ProductSymmetry` one: negating a bit-packed charge is not the component-wise group inverse (negating the encoding of `(1, 2)` decodes as `(-1, -3)`), and `_from_blocks_unchecked` suppresses the validation that would have caught it. That is the same class of defect as **#799 item 2**, where a non-canonical block key passes `_validate` and then `todense()`s to zero — both are hand-rolled charge arithmetic that a `sym.dual` call would get right, which is exactly the boundary #734 was built to establish. Worth fixing as one pass rather than three.
